clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -268,14 +277,17 @@ private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime,
             // A newly created batch can always host the first message.
             if (!batch.tryAppendForSplit(record.timestamp(), record.key(), record.value(), record.headers(), thunk)) {
                 batches.add(batch);
+                addChildrenProducerBatch(batch);
                 batch = createBatchOffAccumulatorForRecord(record, splitBatchSize);
                 batch.tryAppendForSplit(record.timestamp(), record.key(), record.value(), record.headers(), thunk);
             }
         }
 
         // Close the last batch and add it to the batch list after split.
-        if (batch != null)
+        if (batch != null) {
             batches.add(batch);
+            addChildrenProducerBatch(batch);
```

Comment:
The ordering here is critical since we need to make sure the children batches are constructed before we can lifting the parent's future below. I'm wondering if JIT would ever reorder it and thus should we make `childrenProducerBatch` volatile?

Reply:
Is `childrenProducerBatch` ever accessed without a lock?

Reply:
Hello, @guozhangwang not sure if the JIT would ever reorder it, I can change the ` List<ProducerBatch> childrenProducerBatch` to `volatile`, however, making the `List` `volatile` means that the `List` itself will always be read fresh and never thread cached, but the`ProducerBatch` within the `List` will not be, is that what we are looking for? if yes, I can update it for sure.

Reply:
Hello @ijuma although all methods that interact with `childrenProducerBatch` are `synchronized`, I notice now that the method `private synchronized List<ProducerBatch> getChildrenProducerBatch()` returns a reference to the `List` `childrenProducerBatch`, hence two different threads could interact with it without waiting for each other to finish the interaction with the `List childrenProducerBatch`

I will make `synchronized` the method `private void waitForPossibleSplittedBatches()` which is the only place where the method  `getChildrenProducerBatch()` is called, in that way, the access to the  `List` `childrenProducerBatch` will be always `synchronized`.

Let me know please how that sounds to you, please.